debian/nlookup.1
.TH NLOOKUP "1" "April 2003"
.SH NAME
Bnlookup \- Nodelist search tool for bforce
.SH SYNOPSYS
\fBnlookup [-rh]\fR

.SH DESCRIPTION
\fBnlookup\fP is a nodelist search tool for BinkleyForce FTN mailer. 
.SH OPTIONS
\fB\-r\fR
show nodelist string
.P
\fB\-m\fR
show sysops e-mail (see FSP-1004.001 - p2)
.P
\fB-h\fR
show help message
.SH SEE ALSO
bforce(1),bfstat(1),nlookup(1)
.SH AUTHOR
Bforce was written by Alexander Belkin <adb@newmail.ru>, 2:5020/2120
.P
This manual page was written by Zhenja Kaluta <trinfo@mail.ru>
.P
Revision 2 by Evgeniy Kozhuhovskiy <e.kozhuhovskiy@gmail.com>

This is free documentation; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.


CHANGES
* New: now "bforce" can start outgoing sessions simply on stdin/stdout.
  (command line option "-o")
* New: file boxes support. New config keywords "filebox" and
  "filebox_directory"
* Bugfix: now "bforce" will remove .bsy locks immediately after
  disconnection
* New: implemented CRAM-MD5 authentication in the BinkP protocol
* New: new command line option "-c" in the "bfstat" utility
* New: BinkP seems to be working fine now
* Bugfix: some file names could be spoiled on the Hydra protocol while
  receiving (it is a bug with receiving files with the name "d" instead
  of the real one)
* New: implementation of the YooHoo handshake protocol
* New: now will create .csy locks while dialing and convert it to
  the .bsy only at handshake
* Bugfix: now daemon will not call to the nodes marked as Hold/Down
  in a nodelist
* New: added sessions history loging. New config keyword "sessions_history"
  For the history file format look sources.
* Bugfix: Hydra didn't send "short" file name, as a result some mailers
  with strict checks could not receive files with long names :))
* New: removed check for "short" file name length in the Hydra

0.22 (31-07-2000)

* Bugfix: at incoming sessions used speed limit for outgoing sessions
* Bugfix: RH1 mode didn't work correctly
* Bugfix: outgoing BinkP password protected sessions allways fail
  with invalid password error
* Bugfix: files from the AmigaDos Outbound send in incorrect order
* New: passwords are case insensitive now (again)
* New: improved EMSI handshake, now it should work better with systems
  using Cisco or something like that with login prompt
* New: added writing of the remote intro to log (like *ico). To control
  this feature added new option "[No]Intro"
* New: protocol selection options now works at incoming sessions too
* New: added support for modem chat commands, e.g "ATz| ATm0|", mailer
  will wait for modem response ("OK" or "ERROR") after sending each
  substring and only then send the next one
* Bugfix: file masks checker didn't work correctly than '!' (not)
  prefix was used. Now "DelayFiles", "SkipFiles" should work normaly
* New: Added new expression elements: Exist <file name>, Mailer <mailer
  name substring>, Port <port name substring>
* New: Added autoconfiguration of UUCP lock files directory
* Bugfix: daemon could call even to "-Unpublished-" phone if it was
  specified in override.
* Bugfix: could not parse received EMSI data packet if character ']' was
  escaped as '\5d' by remote mailer (e.g. DVMMail do such escaping)
* New: added support for Tx/Rx windows and ASC/UUE packets in the Hydra
* Bugfix: files skipped by remote mailer was not really skipped :)
* New: extended algorithm of renaming received file if file with such name
  allready exist in the inbound directory:
  - For netmail packets and '.tic' files: generate new random name
  - For arcmail packets: rotate extension characters, as most other mailers do
  - For file requests: allways overwrite existing file 
  - For all other files: add numeric extension as in previous versions
* New: added ability to recode file names and intro text to another charset
  then sending/receiving. Recode tables are compartible with HPT. New
  config keywords: "Recode_file_in", "Recode_file_out", "Recode_intro_in"



source/bfutil/nlookup.c
/*
 *	binkleyforce -- unix FTN mailer project
 *	
 *	Copyright (c) 1998-2000 Alexander Belkin, 2:5020/1398.11
 *	
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *	
 *	$Id$
 */

#include "includes.h"
#include "confread.h"
#include "logger.h"
#include "util.h"
#include "nodelist.h"
#define DEF_DOMAIN "fidonet.org"
/*
 *  Our fake expressions checker. Allways return FALSE (?)
 */
bool eventexpr(s_expr *expr)
{
	return FALSE;
}

static void usage(void)
{
	printf_usage("nodelist lookup utility",
		"usage: nlookup [-rmh] <address>\n"
		"\n"
		"options:\n"
		"  -r                show nodelist string\n"
		"  -m                show email address\n"
		"  -h                show this help message\n"
		"\n"
	);
}

void print_nodemail(const s_node *node)
{
	char abuf[BF_MAXADDRSTR+1];

	if( node->sysop && *node->sysop && strcmp(node->sysop, "<none>") )
	{
		char username[BNI_MAXSYSOP+1];
		
		strnxcpy(username, node->sysop, sizeof(node->sysop));
		string_replchar(username, ' ', '_');

		printf("%s@%s.fidonet.org\n", username,
			    ftn_addrstr_inet(abuf, node->addr));
	}
	
	fflush(stdout);
}

void print_nodeinfo(const s_node *node)
{
	char abuf[BF_MAXADDRSTR+1];
	char status[6]; /* Maximum size for node keyword, see
			 * nodelist.c for details */
	
	printf("Address   : %s\n", ftn_addrstr(abuf, node->addr));
	printf("System    : %s\n", node->name);
	printf("Phone     : %s\n", node->phone);
	printf("Sysop     : %s\n", node->sysop);
	printf("Location  : %s\n", node->location);
	printf("Speed     : %lu\n", node->speed);
	printf("Flags     : %s\n", node->flags);

	if( node->worktime.num )
	{
		char timebuf[80];
		time_t unixtime = time(NULL);
		struct tm *now  = localtime(&unixtime);

		timevec_string(timebuf, &node->worktime, sizeof(timebuf));
		
		printf("Work time : %s (%s)\n", timebuf,
			timevec_check(&node->worktime, now) ? "false" : "true");
	}
	
	if( node->sysop && *node->sysop && strcmp(node->sysop, "<none>") )
	{
		char username[BNI_MAXSYSOP+1];
	
		strnxcpy(username, node->sysop, sizeof(node->sysop));
		string_replchar(username, ' ', '_');
		
		printf("E-mail    : %s@%s\n", username,
			ftn_addrstr_inet(abuf, node->addr));
	}
	switch( node->keyword ) {
	case KEYWORD_HOLD :
	     strcpy(status,"Hold");
	     break;
	case KEYWORD_DOWN :
	     strcpy(status,"Down");
	     break;
	case KEYWORD_HUB :
	     strcpy(status,"Hub");
	     break;
	case KEYWORD_HOST :
	     strcpy(status,"Host");
	     break;
	case KEYWORD_PVT :
	     strcpy(status,"Pvt");
	     break;
	case KEYWORD_BOSS :
	     strcpy(status,"Boss");
	     break;
	case KEYWORD_POINT :
	     strcpy(status,"Point");
	     break;
	default:
	     strcpy(status,"Node");
	     break;
	}
	printf("Type      : %s\n",status);
	fflush(stdout);
}

int main(int argc, char *argv[])
{
	s_node node;
	s_faddr addr;
	char ch;
	bool rawstring = FALSE;
	bool emailaddr = FALSE;
	
	/* Initialise random number generation */
	(void)srand((unsigned)time(0));
	/* Initialise current locale */
	(void)setlocale(LC_ALL, "");
	
	while( (ch=getopt(argc, argv, "hrm")) != EOF )
	{
		switch( ch ) {
		case 'h':
			usage();
			exit(BFERR_NOERROR);
		case 'r':
			rawstring = TRUE;
			break;
		case 'm':
			emailaddr = TRUE;
			break;
		default:
			usage();
			exit(BFERR_FATALERROR);
		}
	}

	if( optind >= argc || ftn_addrparse(&addr, argv[optind], FALSE) )
	{
		usage();
		exit(BFERR_FATALERROR);
	}

	if( conf_readconf(conf_getconfname(), 0) )
		exit(BFERR_FATALERROR);
	
	if( rawstring )
	{
		char buf[512];

		if( nodelist_lookup_string(buf, sizeof(buf), addr) == 0 )
			printf("%s\n", buf);
	}
	else if( nodelist_lookup(&node, addr) == 0 )
	{
	    if( emailaddr )
		print_nodemail(&node);
	    else
		print_nodeinfo(&node);
	}
	
	deinit_conf();

	exit(0);
}
